src/mlpack/methods/decision_tree/mse_gain.hpp

```diff
@@ -0,0 +1,152 @@
+   * representation issues. Thus if you are checking for perfect fit, be
+   * sure to use 'gain >= 0.0' and not 'gain == 0.0'. The values vector
+   * should always be of type arma::Row<double> or arma::rowvec.
+   *
+   * @param values Set of values to evaluate MAD gain on.
+   * @param weights Weights associated to each value.
+   * @param begin Start index.
+   * @param end End index.
+   */
+  template<bool UseWeights, typename WeightVecType>
+  static double Evaluate(const arma::rowvec& values,
+                         const WeightVecType& weights,
+                         const size_t begin,
+                         const size_t end)
+  {
+    double mse = 0.0;
+
+    if (UseWeights)
+    {
+      double accWeights = 0.0;
+      double weightedMean = 0.0;
+      WeightedSum(values, weights, begin, end, accWeights, weightedMean);
+
+      // Catch edge case: if there are no weights, the impurity is zero.
+      if (accWeights == 0.0)
+        return 0.0;
+
+      weightedMean /= accWeights;
+
+      for (size_t i = begin; i < end; ++i)
+        mse += weights[i] * std::pow(values[i] - weightedMean, 2);
+
+      mse /= accWeights;
+    }
+    else
+    {
+      double mean = 0.0;
+      Sum(values, begin, end, mean);
+      mean /= (double) (end - begin);
+
+      mse = arma::accu(arma::square(values.subvec(begin, end - 1) - mean));
+      mse /= (double) (end - begin);
+    }
+
+    return -mse;
+  }
+
+  /**
+   * Evaluate the MSE gain on the complete vector.
+   *
+   * @param values Set of values to evaluate MSE gain on.
+   * @param weights Weights associated to each value.
+   */
+  template<bool UseWeights, typename WeightVecType>
+  static double Evaluate(const arma::rowvec& values,
+                         const WeightVecType& weights)
+  {
+    // Corner case: if there are no elements, the impurity is zero.
+    if (values.n_elem == 0)
+      return 0.0;
+
+    return Evaluate<UseWeights>(values, weights, 0, values.n_elem);
+  }
+
+  /**
+   * Calculates the weighted mean squared error gain given the sum of squares
+   * and mean.
+   *
+   * X = array of values of size n.
+   * W = array of weights of size n.
+   *
+   * @f{eqnarray*}{
+   *    MSE = \sum\limits_{i=1}^n {W_i * {X_i}^2} -
+   *        {\dfrac{\sum\limits_{j=1}^n W_j * X_j}
+   *        {\sum\limits_{j=1}^n W_i}}^2
+   * @f}
+   *
+   * @param weightedSumSquares Precomputed weighted sum of square
+   *    (sum(Wi * Xi^2)) of values.
+   * @param weightedMean Precomputed weighted mean (sum(Wi * Xi) / sum(Wi)) of
+   *    values.
+   * @param totalChildWeight Total weight of all the samples in that child.
+   */
+  static double Evaluate(const double weightedSumSquares,
+                         const double weightedMean,
+                         const double totalChildWeight)
+  {
+    double mse = weightedSumSquares / totalChildWeight -
+        weightedMean * weightedMean;
+    return -mse;
+  }
```

Comment:
Yeah! This sounds like a great idea. We can do this. This will remove the necessity of creating multiple overloads for each fitness function.

Reply:
@rcurtin, What do you mean by "incremental" interface?

Also, correct me If I understood you correctly, all we need to do is take the part where we are calculating the best split point into the FitnessFunction classes. Then have two overloads for numeric splits---one with optimizations and the other without optimization. Right?

Also, are we sure that the signatures for these functions that will find the best split efficiently will all require the same kind of input parameters? Combination of all splitting strategies and fitness functions that we have till now, they all require the same set of parameters i.e. data, responses and weights.
But, is there any other splitting strategy or fitness function which you think might be out there that might need anything else?

Reply:
Right, so if we can compute the gain for the split at index `i` based on what the left and right gains were at index `i - 1`, then this would lead to an 'incremental' interface (that might not be the best word for it).  In this scheme, the gain computation function would take in the data points, the index to split on, and then the left and right gains at the previous split.  We would have to establish a unified interface for something like this which provided all the necessary information to a split function: while the left and right gains at index `i - 1` would be sufficient for MSE, would that information also be sufficient for other gain functions?  Or would there be something else we should pass too?  (I didn't consider weights in that discussion, so that may be another parameter that needs to be passed.  There are many variations on the general idea above that could be explored.)

I suppose it would be okay for now to pass the things that MSE needs, and then make sure that the MSE interface can store state somehow (so, e.g., the gain computation is not a `static` function, so that we can store things like the left side's statistics and right side's statistics).  If we get the interface wrong and more needs to be added later, we can always take care of that later.

Anyway, we can check for the presence of this more efficient overload using SFINAE (just like you were investigating), and then choose the correct one from the `SplitIfBetter()` function.

I think I answered everything here... let me know if not. :)

Reply:
Thanks for the explanation. Now, I get the idea. So, I think in general any such optimization has the following structure:-
1. Precompute some data and cache it.
2. Loop through all the points and use the data at the current index to update the cached data for that particular iteration.
3. Use the updated data to compute the gain.

I think we can also have these three things inside our fitness functions. And each of them can implement these interfaces in their own ways. The MSEGain class already has the third part done. I think we can do the other two things as well.

One question I have in my mind is where shall I store the cached data? I was thinking probably it will be good to store it in the FitnessFunction class, but as you mentioned something about "static function", can you tell how will it affect the existing functions?

I also think that this might entirely remove the need to have an additional overload for SplitIfBetter because let's say if a particular FitnessFunction is not optimized, then it can simply have the interfaces (for points 1 and 2 explained above) that do nothing.

Reply:
I agree with what you've written up.  I think we have to store the data as a class member inside the `FitnessFunction`, and then use an instantiated version of it that can hold the cached data.